extractors/reddit/reddit.go

```diff
@@ -0,0 +1,187 @@
+package reddit
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/iawia002/lux/extractors"
+	"github.com/iawia002/lux/request"
+	"github.com/iawia002/lux/utils"
+	"github.com/pkg/errors"
+)
+
+var (
+	redditMP4API = "https://v.redd.it/"
+	redditIMGAPI = "https://i.redd.it/"
+	audioURLPart = "/DASH_audio.mp4"
+	resURLParts  = []string{"/DASH_720.mp4", "/DASH_480.mp4", "/DASH_360.mp4", "/DASH_240.mp4", "/DASH_220.mp4"}
+	res          = []string{"720p", "480p", "360p", "240p", "220p"}
+)
+
+type extractor struct{}
+
+const referer = "https://www.reddit.com"
+
+func init() {
+	extractors.Register("reddit", New())
+}
+
+func New() extractors.Extractor {
+	return &extractor{}
+}
+
+func (e *extractor) Extract(url string, option extractors.Options) ([]*extractors.Data, error) {
+	html, err := request.Get(url, referer, nil)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	// set thread number to 1 manually to avoid http 412 error
+	option.ThreadNumber = 1
+
+	var fileType = ""
+	contentName := utils.MatchOneOf(html, `<title>(.+?)<\/title>`)[1]
+	if utils.MatchOneOf(html, `meta property="og:video" content=.*HLSPlaylist`) != nil {
+		fileType = "mp4"
+	} else if utils.MatchOneOf(html, `<meta property="og:type" content="image"/>`) != nil {
+		fileType = "img"
+	} else if utils.MatchOneOf(html, `https:\/\/preview\.redd\.it\/.*gif`) != nil {
+		fileType = "gif"
+	}
+
+	if fileType == "mp4" {
+		mp4URL := utils.MatchOneOf(html, `https://v.redd.it/(.+?)/HLSPlaylist`)[1]
+		if mp4URL == "" {
+			return nil, errors.New("can't match mp4 content downloadable url")
+		}
+
+		streams := make(map[string]*extractors.Stream, len(resURLParts))
+		for id := 0; id < len(resURLParts); id++ {
+			index := strconv.Itoa(id)
+			id, err := strconv.Atoi(index)
+			if err != nil {
+				return nil, err
+			}
+			resURL := fmt.Sprintf("%s%s%s", redditMP4API, mp4URL, resURLParts[id])
+			audioURL := fmt.Sprintf("%s%s%s", redditMP4API, mp4URL, audioURLPart)
```

Comment:
If all videos share one audio file, then you can generate a separate audio part without having to do it inside the loop